### PR TITLE
HPCC-13963 Allow password reset when required

### PR DESCRIPTION
**Pull request by RussWhitehead**
If an admin sets the "User must reset password at next login" attribute,
LDAP Security Manager needs to be able to detect this and force the Change
Password screen to appear.

Signed-off-by: Russ Whitehead william.whitehead@lexisnexis.com

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@wangkx Please review

**Comment by wangkx** [comment]
@RussWhitehead looks good.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge and close

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 1]
```diff
@@ -2608,7 +2613,7 @@ class CLdapClient : public CInterface, implements ILdapClient
         StringBuffer filter;
         filter.append("sAMAccountName=").append(username);
 
-        char        *attrs[] = {"cn", NULL};
+        char        *attrs[] = {"distinguishedName", NULL};
         CLDAPMessage searchResult;
```
Comment (richardkchapman) [inline system/security/LdapSecurity/ldapconnection.cpp]: What's the rationale for this change, and what will it affect?

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 2]
```diff
@@ -2627,11 +2632,11 @@ class CLdapClient : public CInterface, implements ILdapClient
                   attribute != NULL;
                   attribute = atts.getNext())
             {
-                if(0 == stricmp(attribute, "cn"))
+                if(0 == stricmp(attribute, "distinguishedName"))
                 {
                     CLDAPGetValuesWrapper vals(ld, message, attribute);
                     if (vals.hasValues())
-                        userdn.append("cn=").append(vals.queryValues()[0]).append(",").append(m_ldapconfig->getUserBasedn());
+                        userdn.set(vals.queryValues()[0]);
```
Comment (richardkchapman) [inline system/security/LdapSecurity/ldapconnection.cpp]: What's the rationale for this change, and what will it affect?

**Comment by richardkchapman** [comment]
@RussWhitehead A couple of questions

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 2 above)
Comment (RussWhitehead) [inline system/security/LdapSecurity/ldapconnection.cpp]: The following call to ldap_modify_s() requires a fully qualified DN distinguished name (ie ou=wwhitehead,ou=users,ou=ecl,DC=RISKDev,DC=LOCAL). In the legacy code, this was built up from the "usersBaseDN" setting in the config, followed by the "cn" that was queried before this change.  Because now we support nested users (ie they reside somewhere below the "users"), I need to know the fully qualified DN, including the levels below "users".  I tested this thoroughly in both legacy and RISKDev environments.

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 1 above)
Comment (RussWhitehead) [inline system/security/LdapSecurity/ldapconnection.cpp]: See explanation below, no need to build up the DN when we can just query it
